### ISSUE TITLE
Attempt to raise the soft limit

### ISSUE THREAD

**Issue by ekohl** [eclipse/mosquitto#2674]
I'm forwarding this issue from https://bugzilla.redhat.com/show_bug.cgi?id=2132774.

The environment is that Mosquitto 2.0.14 is started via systemd on RHEL8.

It's observed that at most 1015 connections are accepted by default, but if a systemd override with `LimitNOFILE` is made it can be increased further. This surprised me because systemd on EL8 by default should give a hard limit of 262144 but has a soft limit of 1024. The 1015 connections are surprisingly close to 1024. The 9 other ones can probably be accounted for by stdin/stdout/stderr and some other files like configs.

https://github.com/systemd/systemd/blob/23f3a6f5ff864fd26063c6c35fdaa6d85de566c7/NEWS#L5880-L5914 does have quite a bit to say on this. 

> Note that the soft limit remains at 1024 for compatibility reasons: the
          traditional UNIX select() call cannot deal with file descriptors >=
          1024 and increasing the soft limit globally might thus result in
          programs unexpectedly allocating a high file descriptor and thus
          failing abnormally when attempting to use it with select() (of
          course, programs shouldn't use select() anymore, and prefer
          poll()/epoll, but the call unfortunately remains undeservedly popular
          at this time).

Then later:

> Programs that want to take benefit of the increased limit have to "opt-in" into
          high file descriptors explicitly by raising their soft limit.

So my question is: should Mosquitto attempt to raise its soft limit to, by default, get many more file descriptors or is this unsafe due to `select()` usage?

**Comment by rswindell** [comment]
`select()` will still have the 1024 socket limit (at least on Linux):
`/usr/include/linux/posix_types.h:#define __FD_SETSIZE   1024`

Mosquitto should be changed to use `poll()` instead.

**Comment by ekohl** [comment]
To be clear: I don't know if mosquitto uses `select()`. I'm not a C programmer and approaching this from a user point of view.

**Comment by ekohl** [comment]
Someone pointed me to https://github.com/eclipse/mosquitto/pull/1884 which may solve it.

**Comment by pablomh** [comment]
While we wait for the long term solution, could we add an override systemd unit file like the one shown in the BZ?

I'm testing right now one with `LimitNOFILE=infinity` that after https://github.com/systemd/systemd/pull/5795 will default its value to the max open files set by the kernel.

**Comment by ekohl** [comment]
@pablomh any luck with `LimitNOFILE=infinity`?

**Comment by ralight** [comment]
Something like `LimitNOFILE=16384` works fine. The mosquitto broker doesn't use select so there's no other inbuilt limit.

**Comment by pablomh** [comment]
Yes, @ekohl. We've been been using it for a while in our scale testing with no issues that have been noticed.

**Comment by ralight** [comment]
Closing as fixed

**Comment by ekohl** [comment]
@ralight I tried to look through the git commits but I see you've been very busy (yay). Can you point me to the commit and/or PR that addressed it?

**Comment by ralight** [comment]
@ekohl It is the comments above about using e.g. `LimitNOFILE=16384` in the systemd unit file, no code changes were required.

**Comment by ekohl** [comment]
I think it would be clearer if this was closed as not completed instead of completed because:

* I don't see a change in https://github.com/eclipse-mosquitto/mosquitto/tree/master/service/systemd to raise the default
* There is no code to increase the soft limit to benefit from systemd's increased default hard limits (potentially guarded behind `WITH_SYSTEMD`)
* https://github.com/eclipse-mosquitto/mosquitto/pull/1884 is still open

**Comment by ralight** [comment]
This limit is something that should be set by the end user. The default value is more than adequate for most cases, it would be entirely inappropriate to increase the limit for everyone.

Increasing the limit in systemd works exactly by changing LimitNOFILE, there are no code changes required.

The linked issue is completely unrelated to the broker.

This issue is closed as completed because the question of how to increase the number of sockets available to the broker when running under systemd is answered.

**Comment by ekohl** [comment]
> This issue is closed as completed because the question of how to increase the number of sockets available to the broker when running under systemd is answered.

Ah, it wasn't clear to me that it was unrelated and that clarification helps.

I was looking at https://mosquitto.org/documentation/ but I don't see an obvious place where I could contribute anything. Is it something that would make sense as a comment in the systemd service files? Or in short, how can a user figure this out without reading GitHub issues?
